tc-messaging/src/main/java/com/tc/bytes/TCReference.java

```diff
@@ -0,0 +1,32 @@
+/*
+ *
+ *  The contents of this file are subject to the Terracotta Public License Version
+ *  2.0 (the "License"); You may not use this file except in compliance with the
+ *  License. You may obtain a copy of the License at
+ *
+ *  http://terracotta.org/legal/terracotta-public-license.
+ *
+ *  Software distributed under the License is distributed on an "AS IS" basis,
+ *  WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for
+ *  the specific language governing rights and limitations under the License.
+ *
+ *  The Covered Software is Terracotta Core.
+ *
+ *  The Initial Developer of the Covered Software is
+ *  Terracotta, Inc., a Software AG company
+ *
+ */
+package com.tc.bytes;
+
+/**
+ *
+ */
+public interface TCReference extends Iterable<TCByteBuffer>, AutoCloseable {
```

Comment:
It's kinda bothering me that we're exposing both the array and the iterable... any way we could standardize on one of the other? I'm thinking that would simplify some of the implementations and prevent the need for conversion back and forth?

Reply:
I'd like to get rid of the array stuff in a gradual way.  maybe depreciate asArray?

Reply:
Hmm... I was thinking of going the other way (maybe). The only thing tied heavily to the iterator right now is the read logic in the input stream that calls `iterator.remove()`. If you moved that inside the `TCReference` it might be easier to the array way and move to arrays for everything?

Reply:
Actually, it looks like I culled most of the places asArray is used.  I'll try and remove it.

Reply:
yeah.  It can be removed but I will do that in the adoption PR.

Reply:
oh man.  Dealing with TCByteBuffer[] in the codebase is a total pain.  In a total contradiction to my earlier comment, I'd much rather have TCByteBuffer[] start migrating to TCReference.

Reply:
I think it will be more clear when I push the adoption PR.